// lightning-base-components/src/lightning/routingService/routingService.js
import formFactor from '@salesforce/client/formFactor';
import { shouldDispatchToApp } from './routingServiceUtils';

const GET_LINK_INFO_EVENT = 'lightningroutingservicegetlinkinfo';

export const urlTypes = {
    standard: 'standard_webPage',
};

export class LinkInfo {
    constructor(url, dispatcher) {
        this.url = url;
        this.dispatcher = dispatcher;
        Object.freeze(this);
    }
}

export function registerLinkProvider(element, providerFn) {
    element.addEventListener(GET_LINK_INFO_EVENT, providerFn);
}

export function unregisterLinkProvider(element, providerFn) {
    element.removeEventListener(GET_LINK_INFO_EVENT, providerFn);
}

/*
 * Mock getLinkInfo
 *
 * @returns {Promise[LinkInfo]}
 */
export function getLinkInfo(element, stateRef) {
    return new Promise((resolve, reject) => {
        // eslint-disable-next-line lightning-global/no-custom-event-identifier-arguments
        const getLinkInfoEvent = new CustomEvent(GET_LINK_INFO_EVENT, {
            detail: {
                stateRef,
                callback: (err, linkInfo) => {
                    if (err) {
                        reject(err);
                    } else {
                        resolve(linkInfo);
                    }
                },
            },
            bubbles: true,
            composed: true,
            cancelable: true,
        });
        element.dispatchEvent(getLinkInfoEvent);
    });
}

/**
 * Determines the route for the given url and updates the element
 * state with the correct url and dispatcher.
 *
 * @param {HTMLElement} element Element from which to dispatch the routing event
 * @param {Object} url Link to route, target Target of the link
 * @param {function} callback on the returned LinkInfo
 *
 * @returns {Promise} Promise[LinkInfo]
 */
export function updateRawLinkInfo(element, { url, target }) {
    if (url === undefined || url === null) {
        // eslint-disable-next-line no-console
        console.error('url must be specified');
    }

    const dispatchToApp = shouldDispatchToApp(target, formFactor);

    if (dispatchToApp) {
        return getLinkInfo(element, {
            stateType: urlTypes.standard,
            attributes: {
                url,
                target,
            },
        });
    }

    // Return a no-op dispatcher for targets that should be handled by the browser
    return new Promise((resolve) => {
        resolve({ url, dispatcher: () => {} });
    });
}
